hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerSelectionCriteria.java

```diff
@@ -88,19 +93,25 @@ private boolean isContainerReplicatingOrDeleting(ContainerID containerID) {
    * 5. Container should be closed.
    * 6. If the {@link LegacyReplicationManager} is enabled, then the container should not be an EC container.
    * @param node DatanodeDetails for which to find candidate containers.
-   * @return NavigableSet of candidate containers that satisfy the criteria.
+   * @return Set of candidate containers that satisfy the criteria.
    */
-  public NavigableSet<ContainerID> getCandidateContainers(
+  public Set<ContainerID> getCandidateContainers(
       DatanodeDetails node, long sizeMovedAlready) {
-    NavigableSet<ContainerID> containerIDSet =
-        new TreeSet<>(orderContainersByUsedBytes().reversed());
     try {
-      containerIDSet.addAll(nodeManager.getContainers(node));
+      // Initialize containerSet for node
+      if (!setMap.containsKey(node)) {
+        addNodeToSetMap(node);
+      }
+      // In case the node is removed
+      nodeManager.getContainers(node);
```

Comment:
Are we making this call just to see if an exception gets thrown? In that case this is a bit awkward and confusing. Does node manager provide an API that we can use first to check if SCM knows the node, and then get its containers (or remove them from the cache is the node isn't there anymore)?

Reply:
Yes, since currently there is no explicit method to check if a node exists, this is used for this check.

Reply:
I agree with @sodonnel.  Would `isNodeRegistered` be even better?

https://github.com/apache/ozone/blob/ac597ad05985641faa51500a80fa46a6b5608723/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java#L627-L634